massa-execution-worker/src/context.rs

```diff
@@ -0,0 +1,408 @@
+    /// * previous_changes: list of ledger changes that happened since the final ledger state and before the current execution
+    ///
+    /// # returns
+    /// A new (empty) ExecutionContext instance
+    pub(crate) fn new(
+        final_ledger: Arc<RwLock<FinalLedger>>,
+        previous_changes: LedgerChanges,
+    ) -> Self {
+        ExecutionContext {
+            speculative_ledger: SpeculativeLedger::new(final_ledger, previous_changes),
+            max_gas: Default::default(),
+            gas_price: Default::default(),
+            slot: Slot::new(0, 0),
+            created_addr_index: Default::default(),
+            created_event_index: Default::default(),
+            opt_block_id: Default::default(),
+            stack: Default::default(),
+            read_only: Default::default(),
+            events: Default::default(),
+            unsafe_rng: Xoshiro256PlusPlus::from_seed([0u8; 32]),
+            origin_operation_id: Default::default(),
+        }
+    }
+
+    /// Returns a snapshot containing the clone of the current execution state.
+    /// Note that the snapshot does not include slot-level information such as the slot number or block ID.
+    pub(crate) fn get_snapshot(&self) -> ExecutionContextSnapshot {
+        ExecutionContextSnapshot {
+            ledger_changes: self.speculative_ledger.get_snapshot(),
+            created_addr_index: self.created_addr_index,
+            created_event_index: self.created_event_index,
+            stack: self.stack.clone(),
+            events: self.events.clone(),
+            unsafe_rng: self.unsafe_rng.clone(),
+        }
+    }
+
+    /// Resets context to an existing snapshot
+    /// Note that the snapshot does not include slot-level information such as the slot number or block ID.
+    pub fn reset_to_snapshot(&mut self, snapshot: ExecutionContextSnapshot) {
+        self.speculative_ledger
+            .reset_to_snapshot(snapshot.ledger_changes);
+        self.created_addr_index = snapshot.created_addr_index;
+        self.created_event_index = snapshot.created_event_index;
+        self.stack = snapshot.stack;
+        self.events = snapshot.events;
+        self.unsafe_rng = snapshot.unsafe_rng;
+    }
+
+    /// Create a new ExecutionContext for readonly execution
+    /// This should be used before performing a readonly execution.
+    ///
+    /// # arguments
+    /// * slot: slot at which the execution will happen
+    /// * req: parameters of the read only execution
+    /// * previous_changes: list of ledger changes that happened since the final ledger state and before this execution
+    /// * final_ledger: thread-safe access to the final ledger. Note that this will be used only for reading, never for writing
+    ///
+    /// # returns
+    /// A ExecutionContext instance ready for a read-only execution
+    pub(crate) fn new_readonly(
```

Comment:
the usual practice is to elude the redundant part because you have to write the type:  `HashMap::with_capacity()` instead of `HashMap::new_hashmap_with_capacity`.

It looks like this in practice: `ExecutionContext::new_readonly()`

Reply:
I guess that works :) Done